### ISSUE TITLE
Explore tools for annotating intermediate layers of ML models with range bounds

### ISSUE THREAD

**Issue by j2kun** [google/heir#1700]
From the meeting notes on parameter selection https://docs.google.com/document/d/1ASmm8UiQisMyk1EQYVMSw--I095nsTswZEPJ6BlcrIQ/edit?usp=sharing

There seem to be two approaches:

## Runtime analysis using a validation set

For these, most techniques are ad-hoc. One uses something like pytorch's [register_module_forward_hook](https://pytorch.org/docs/stable/generated/torch.nn.modules.module.register_module_forward_hook.html) to dump intermediate values, and run it on a validation set to get the values dumped to some format.

We can do this and parse the results back to annotations on the pytorch model, and try to get those preserved when dumping to MLIR. A colleague of mine at Google is going to tinker with a nice way to do this, and next week I can try to convert his work to produce MLIR annotations.

## Formal methods

There are tools like https://github.com/Verified-Intelligence/alpha-beta-CROWN (this one seems the best and most actively developed) and https://github.com/vas-group-imperial/VeriNet that are purely based on static analysis and provide interval-based propagation through a model for worst-case bounds.

I haven't seen these tools before today, but alpha-beta-CROWN seems to support pytorch and ONNX and there is a usage doc here: https://github.com/Verified-Intelligence/alpha-beta-CROWN/blob/main/complete_verifier/docs/abcrown_usage.md

**Comment by j2kun** [comment]
CC @ZenithalHourlyRate maybe you would want to try using alpha-beta-CROWN on some test model in the mean time to see what it outputs?

**Comment by j2kun** [comment]
While working on a potential torch integration, filed https://github.com/pytorch/xla/issues/8993 to ask about attribute propagation.

**Comment by j2kun** [comment]
Here is my prototype for saving intermediate layer values https://github.com/j2kun/torch-intermediate

I have not tried this in conjunction with torch-mlir, just the stableHLO export. https://github.com/pytorch/xla/issues/8993#issuecomment-2813595343 has an input/output example

CC @AlexanderViand

**Comment by AlexanderViand** [comment]
WIP here: https://github.com/llvm/torch-mlir/compare/main...belfortlabs:torch-mlir:preserve-metadata
